### PR TITLE
Upgrade base images, benchmark images, and Python.

### PR DESCRIPTION
**Pull request by Alan32Liu**
I guess this branch has what we need? Not sure why it does not has a PR though.
I make it up-to-date with the master branch so that we can discuss here.

### CODE REVIEW

**Comment by Alan32Liu** [comment]
A quick update:
I am working on merging modifications from other PR to here.
Currently, I have a `docker/base-image/Dockerfile` that is based on `focal` and has `python3.10.8` (the latest version).
I also updated the pip packages in requirements to the latest version and managed to install them successfully in `docker/base-image/Dockerfile`.
I will push later.

**Comment by Alan32Liu** [comment]
TODOs in this PR:
1. Check if there is any conflict with the new version of `Ubuntu`, `python`, and `pip` packages (like #1521).
2. Check if this PR missed anything in #1493.
3. Use a newer OSS-Fuzz docker image (like #1441).
4. Final check for conflicts.

**Comment by Alan32Liu** [comment]
TODOs in future PRs:
1. Fix new lint/formatter issues, as we are using a newer version of `pylint` now.
2. Update documentation.

**Comment by jonathanmetzman** [comment]
btw you can just checkout https://github.com/google/fuzzbench/pull/1441/files locally and start pushing to that PR. I don't think we need another one.

File: docker/image_types.yaml [hunk 1]
```diff
@@ -49,7 +49,6 @@
     - 'parent_image=gcr.io/fuzzbench/builders/coverage/{benchmark}-intermediate'
   depends_on:
     - 'coverage-{benchmark}-builder-intermediate'
-    - 'base-image'
```
Comment (jonathanmetzman) [inline docker/image_types.yaml:52]: What's the reason for this removal?

Comment (Alan32Liu) [reply docker/image_types.yaml:52]: > What's the reason for this removal?

I do not know, it comes with this branch of yours : )

**Comment by Alan32Liu** [comment]
CI failed because we've updated `pylint`.

**Comment by Alan32Liu** [comment]
Fixing the lint errors requires tons of changes (I've fixed 50+ and there are a lot more).
Would it be a good idea to split that into a different PR?

**Comment by Alan32Liu** [comment]
> Fixing the lint errors requires tons of changes (I've fixed 50+ and there are a lot more). Would it be a good idea to split that into a different PR?

Most fo the fixes are trivial (e.g., f-string, specify encoding in `open`).

File: docker/image_types.yaml (hunk 1 above)
Comment (Alan32Liu) [reply docker/image_types.yaml:52]: Would you prefer me to revert this?

**Comment by Alan32Liu** [comment]
I reckon the failure is caused by the outdated python version in [`benchfuzz/Dockerfile`](https://github.com/google/fuzzbench/pull/1526/commits/d9142e8c2242a225fe52a71815d62e2b018406c6#diff-58a2972ec9175b10a7598e60291f1dc8e6671cc6aac9e870c7dd9dc8f7321c4cL15), which will be built as `gcr.io/fuzzbench/base-image`, if I am not mistaken?

**Comment by Alan32Liu** [comment]
Not sure if this is important, but:
1. Image `gcr.io/oss-fuzz-base/base-builder:latest`'s default `python` version is `3.8.3`, it does not have `3.10`.
2. Image `gcr.io/oss-fuzz-base/base-builderr@sha256:fb1a9a49752c9e504687448d1f1a048ec1e062e2e40f7e8a23e86b63ff3dad7c`'s default `python` version is `3.8.10`, it does not have `3.10` either.

File: docker/image_types.yaml (hunk 1 above)
Comment (jonathanmetzman) [reply docker/image_types.yaml:52]: I think this is causing the CI failure as we discussed

**Comment by Alan32Liu** [comment]
Would there be any fuzzer that we'd like to exclude/update?
I noticed that many fuzzers have outdated dependencies, e.g. it seems `lafintel` relies on an old version of `/usr/local/include/c++/v1/` to build, `klee` failed due to `Host Clang must be able to find libstdc++4.8 or newer!`

Would we want to stop maintaining or updating some fuzzers in the new OS?

**Comment by Alan32Liu** [comment]
By the way, I fixed the build issues of some fuzzers by manually installing their old version of dependencies, which dated back to `xenial`(Ubuntu:18) or even `trusty` (Ubuntu:16).
But I feel that if they are actively maintained by their authors, then we probably should update them to the latest version (and update dependencies to adapt to the new OS, if available); if their authors stopped maintaining them, then we probably should do the same?

If a fuzzer has a really fantastic performance, then we probably should keep it even if it is no longer under maintaince.

**Comment by Alan32Liu** [comment]
@jonathanmetzman: Would there be any chance that you could help me with the build failures of `libFuzzer` and `KLEE` (e.g., with `libpng-1.2.56` as a sample benchmark program)?
I have been trying to fix them for a while, but unfortunately, I was not able to make them work 🙁.
Thanks!

**Comment by Alan32Liu** [comment]
Also, many `oss-fuzz` and `bug` CIs failed.
For example, `afl, oss-fuzz` and `afl, bug` failed, even though `afl, standard` passed.
What are the differences between them are how to debug&fix them?

**Comment by Alan32Liu** [comment]
Dumping the `Python3` installation output, which creates too much noise in the logs and makes them hard to load/read.

**Comment by Alan32Liu** [comment]
It is OK to let `presubmit` fail for now: I will bring in the PR that fixes it at the end.

**Comment by Alan32Liu** [comment]
@jonathanmetzman : Please let me know if [Fix test_build_all_measurers](https://github.com/google/fuzzbench/pull/1526/commits/63f0eaa8db663c1bb928ff558893f67d75741c5e) makes sense.
I was unsure if I completely understood the previous code (which failed the unittest), so I was trying to fix it based on its function description.

**Comment by Alan32Liu** [comment]
Not sure what is the best way to fix `validate_fuzzers_and_benchmarks` in `presubmit`:
I guess we don't want to delete the temporary fuzzers and variations now, but also do not want to always check them in CIs.

**Comment by Alan32Liu** [comment]
> Not sure what is the best way to fix `validate_fuzzers_and_benchmarks` in `presubmit`: I guess we don't want to delete the temporary fuzzers and variations now, but also do not want to always check them in CIs.
> 
> That is also the only thing left to pass all CI tests.

@jonathanmetzman : A gentle ping on this : )
Would now be a good time to delete the temporary fuzzers or remove them from CIs, so that we can merge this PR to master?
Also, I noticed that there are experiments running for weeks (`r-2022-11-11-um-prioritize-2209607`) or months (`r-2022-10-13-um-final-3-2201231`), is that intended?
Thanks!

**Comment by Alan32Liu** [comment]
@jonathanmetzman:
Here is the full list of current fuzzers, could you please also tick the ones that you'd like to keep? Thanks!
I will delete the rest.
*  [x] afl
*  [x] aflfast
*  [x] aflplusplus
*  [x] aflsmart
*  [x] centipede
*  [x] eclipser
*  [x] fairfuzz
*  [x] honggfuzz
*  [x] libafl
*  [x] libfuzzer
*  [x] mopt
*  [x] klee               # To Be Fixed.
*  [x] symcc_aflplusplus  # To Be Fixed.
*  [ ] afl_2_52_b
*  [ ] aflcc
*  [ ] afl_collision_free
*  [ ] afl_double_timeout
*  [ ] afl_maxmap
*  [ ] afl_no_favfactor
*  [ ] afl_no_favored
*  [ ] afl_no_trim
*  [ ] aflplusplus_classic
*  [ ] aflplusplus_cmplog
*  [ ] aflplusplus_cmplog_1585
*  [ ] aflplusplus_cmplog_1587
*  [ ] aflplusplus_cmplog_r
*  [ ] aflplusplus_dict2file
*  [ ] aflplusplus_frida
*  [ ] aflplusplus_muttfuzz
*  [ ] aflplusplus_optimal
*  [ ] aflplusplus_pcguard
*  [ ] aflplusplus_qemu
*  [ ] aflplusplus_qemu_tracepc
*  [ ] aflplusplus_tracepc
*  [ ] aflplusplus_um_parallel
*  [ ] aflplusplus_um_prioritize
*  [ ] aflplusplus_um_prioritize_75
*  [ ] aflplusplus_um_random
*  [ ] aflplusplus_um_random_3
*  [ ] aflplusplus_um_random_6
*  [ ] aflplusplus_um_random_75
*  [ ] aflplusplus_zafl
*  [ ] aflpp_random_default
*  [ ] aflpp_random_no_favs
*  [ ] aflpp_random_wrs
*  [ ] aflpp_random_wrs_rf
*  [ ] aflpp_random_wrs_rf_rp
*  [ ] aflpp_random_wrs_rp
*  [ ] afl_qemu
*  [ ] afl_random_favored
*  [ ] afl_scheduling_lifo
*  [ ] afl_scheduling_random
*  [ ] afl_score_max
*  [ ] afl_score_min
*  [ ] afl_score_no_novel_prioritization
*  [ ] afl_score_random
*  [ ] afl_splicing_mutation
*  [ ] afl_two_instances
*  [ ] afl_um_parallel
*  [ ] afl_um_prioritize
*  [ ] afl_um_random
*  [ ] afl_virginmap
*  [ ] centipede_function_filter
*  [ ] coverage
*  [ ] eclipser_aflplusplus
*  [ ] eclipser_um_parallel
*  [ ] eclipser_um_prioritize
*  [ ] eclipser_um_prioritize_75
*  [ ] eclipser_um_random
*  [ ] eclipser_um_random_75
*  [ ] entropic
*  [ ] entropic_execute_final
*  [ ] fafuzz
*  [ ] fuzzolic_aflplusplus_fuzzy
*  [ ] fuzzolic_aflplusplus_z3
*  [ ] glibfuzzer
*  [ ] gramatron
*  [ ] grimoire
*  [ ] honggfuzz_qemu
*  [ ] honggfuzz_um_parallel
*  [ ] honggfuzz_um_prioritize
*  [ ] honggfuzz_um_prioritize_75
*  [ ] honggfuzz_um_random
*  [ ] honggfuzz_um_random_75
*  [ ] introspector_driven_focus
*  [ ] lafintel
*  [ ] libafl_text
*  [ ] libfuzzer_dataflow
*  [ ] libfuzzer_dataflow_load
*  [ ] libfuzzer_dataflow_pre
*  [ ] libfuzzer_dataflow_store
*  [ ] libfuzzer_exeute_final
*  [ ] libfuzzer_focus_idx0
*  [ ] libfuzzer_focus_idx1
*  [ ] libfuzzer_focus_idx2
*  [ ] libfuzzer_focus_idx3
*  [ ] libfuzzer_focus_idx4
*  [ ] libfuzzer_focus_idx5
*  [ ] libfuzzer_focus_idx6
*  [ ] libfuzzer_focus_idx7
*  [ ] libfuzzer_focus_idx8
*  [ ] libfuzzer_fork_parallel
*  [ ] libfuzzer_um_parallel
*  [ ] libfuzzer_um_prioritize
*  [ ] libfuzzer_um_prioritize_75
*  [ ] libfuzzer_um_random
*  [ ] libfuzzer_um_random_75
*  [ ] manul
*  [ ] nautilus
*  [ ] neuzz
*  [ ] pythia_bb
*  [ ] pythia_effect_bb
*  [ ] symcc_afl
*  [ ] symcc_aflplusplus_single
*  [ ] symcc_afl_single
*  [ ] symqemu_aflplusplus
*  [ ] token_level
*  [ ] tortoisefuzz
*  [ ] weizz_qemu
*  [ ] wingfuzz

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman: Here is the full list of current fuzzers, could you please also tick the ones that you'd like to keep? Thanks! I will delete the rest.
> 
> * [x]  afl
> * [x]  aflfast
> * [x]  aflplusplus
> * [x]  aflsmart
> * [x]  centipede
> * [x]  eclipser
> * [x]  fairfuzz
> * [x]  honggfuzz
> * [x]  libafl
> * [x]  libfuzzer
> * [x]  mopt
> * [x]  klee               # To Be Fixed.
> * [x]  symcc_aflplusplus  # To Be Fixed.
> 
> * [ ]  afl_2_52_b
> * [ ]  aflcc
> * [ ]  afl_collision_free
> * [ ]  afl_double_timeout
> * [ ]  afl_maxmap
> * [ ]  afl_no_favfactor
> * [ ]  afl_no_favored
> * [ ]  afl_no_trim
> * [ ]  aflplusplus_classic
> * [ ]  aflplusplus_cmplog
> * [ ]  aflplusplus_cmplog_1585
> * [ ]  aflplusplus_cmplog_1587
> * [ ]  aflplusplus_cmplog_r
> * [ ]  aflplusplus_dict2file
> * [ ]  aflplusplus_frida
> * [ ]  aflplusplus_muttfuzz
> * [ ]  aflplusplus_optimal
> * [ ]  aflplusplus_pcguard
> * [ ]  aflplusplus_qemu
> * [ ]  aflplusplus_qemu_tracepc
> * [ ]  aflplusplus_tracepc
> * [ ]  aflplusplus_um_parallel
> * [ ]  aflplusplus_um_prioritize
> * [ ]  aflplusplus_um_prioritize_75
> * [ ]  aflplusplus_um_random
> * [ ]  aflplusplus_um_random_3
> * [ ]  aflplusplus_um_random_6
> * [ ]  aflplusplus_um_random_75
> * [ ]  aflplusplus_zafl
> * [ ]  aflpp_random_default
> * [ ]  aflpp_random_no_favs
> * [ ]  aflpp_random_wrs
> * [ ]  aflpp_random_wrs_rf
> * [ ]  aflpp_random_wrs_rf_rp
> * [ ]  aflpp_random_wrs_rp
> * [ ]  afl_qemu
> * [ ]  afl_random_favored
> * [ ]  afl_scheduling_lifo
> * [ ]  afl_scheduling_random
> * [ ]  afl_score_max
> * [ ]  afl_score_min
> * [ ]  afl_score_no_novel_prioritization
> * [ ]  afl_score_random
> * [ ]  afl_splicing_mutation
> * [ ]  afl_two_instances
> * [ ]  afl_um_parallel
> * [ ]  afl_um_prioritize
> * [ ]  afl_um_random
> * [ ]  afl_virginmap
> * [ ]  centipede_function_filter
> * [ ]  coverage
> * [ ]  eclipser_aflplusplus
> * [ ]  eclipser_um_parallel
> * [ ]  eclipser_um_prioritize
> * [ ]  eclipser_um_prioritize_75
> * [ ]  eclipser_um_random
> * [ ]  eclipser_um_random_75
> * [ ]  entropic
> * [ ]  entropic_execute_final
> * [ ]  fafuzz
> * [ ]  fuzzolic_aflplusplus_fuzzy
> * [ ]  fuzzolic_aflplusplus_z3
> * [ ]  glibfuzzer
> * [ ]  gramatron
> * [ ]  grimoire
> * [ ]  honggfuzz_qemu
> * [ ]  honggfuzz_um_parallel
> * [ ]  honggfuzz_um_prioritize
> * [ ]  honggfuzz_um_prioritize_75
> * [ ]  honggfuzz_um_random
> * [ ]  honggfuzz_um_random_75
> * [ ]  introspector_driven_focus
> * [ ]  lafintel
> * [ ]  libafl_text
> * [ ]  libfuzzer_dataflow
> * [ ]  libfuzzer_dataflow_load
> * [ ]  libfuzzer_dataflow_pre
> * [ ]  libfuzzer_dataflow_store
> * [ ]  libfuzzer_exeute_final
> * [ ]  libfuzzer_focus_idx0
> * [ ]  libfuzzer_focus_idx1
> * [ ]  libfuzzer_focus_idx2
> * [ ]  libfuzzer_focus_idx3
> * [ ]  libfuzzer_focus_idx4
> * [ ]  libfuzzer_focus_idx5
> * [ ]  libfuzzer_focus_idx6
> * [ ]  libfuzzer_focus_idx7
> * [ ]  libfuzzer_focus_idx8
> * [ ]  libfuzzer_fork_parallel
> * [ ]  libfuzzer_um_parallel
> * [ ]  libfuzzer_um_prioritize
> * [ ]  libfuzzer_um_prioritize_75
> * [ ]  libfuzzer_um_random
> * [ ]  libfuzzer_um_random_75
> * [ ]  manul
> * [ ]  nautilus
> * [ ]  neuzz
> * [ ]  pythia_bb
> * [ ]  pythia_effect_bb
> * [ ]  symcc_afl
> * [ ]  symcc_aflplusplus_single
> * [ ]  symcc_afl_single
> * [ ]  symqemu_aflplusplus
> * [ ]  token_level
> * [ ]  tortoisefuzz
> * [ ]  weizz_qemu
> * [ ]  wingfuzz

We need to keep coverage for sure.
Let's do the upgrade and then delete in a subsequent PR (OK if CI fails)

**Comment by Alan32Liu** [comment]
> We need to keep coverage for sure. Let's do the upgrade and then delete in a subsequent PR (OK if CI fails) 

Sure! I will merge this on Monday, then.

> Explain why we need to delete some of these please. Is it because they need to be fixed to pass linting?

Yep, it is mainly to avoid the `validate_fuzzers_and_benchmarks ` error from `presubmit`.

**Pull request merged**